kubelet: invoke part of image verification policy earlier

atykhyy opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Some decisions about image pull credential verification can be applied without reference to image pull credentials: if the policy is NeverVerify, or if it is NeverVerifyPreloadedImages and the image is preloaded, or if is NeverVerifyAllowListedImages and the image is white-listed. In these cases, there is no need to look up credentials.

#### Which issue(s) this PR is related to:

Related PR: #133079

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

No.

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

pacoxu commented:
/release-note-none
/ok-to-test 
/lgtm
/assign @stlaz @liggitt

pacoxu commented:
/lgtm cancel
UT needs an update.

atykhyy commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -178,7 +178,8 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, objRef *v1.ObjectR
 
-	if imageRef != "" && !utilfeature.DefaultFeatureGate.Enabled(features.KubeletEnsureSecretPulledImages) {
+	// run the part of image verification policy that does not depend on credentials before looking them up
+	if imageRef != "" && !(utilfeature.DefaultFeatureGate.Enabled(features.KubeletEnsureSecretPulledImages) && m.imagePullManager.MustAttemptImagePull(requestedImage, imageRef, nil, nil)) {
```
> after the feature goes to GA, this if block can be removed

If this if block is removed, the problem described in #133079 will reappear.

pacoxu replied:
> > after the feature goes to GA, this if block can be removed
> 
> If this if block is removed, the problem described in #133079 will reappear.

Then just fine for now. But it seems that this still needs an update once this is Beta or GA.

atykhyy commented:
/retest-required

stlaz reviewed (changes requested):
I believe we should keep `MustAttemptImagePull()` the single place that evaluates the policies and performs the decision whether to pull based on the underlying cache state.

We can do that by performing the actual credential pull JIT and caching the result.

This PR is also missing unit tests for the fixed behavior.

atykhyy commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -257,6 +245,37 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, objRef *v1.ObjectR
 	return m.pullImage(ctx, logPrefix, objRef, pod.UID, requestedImage, spec, pullCredentials, podSandboxConfig)
 }
 
+func (m *imageManager) getPullCredentials(image string, pod *v1.Pod, pullSecrets []v1.Secret, podSandboxConfig *runtimeapi.PodSandboxConfig) ([]credentialprovider.TrackedAuthConfig, error) {
```
It's used for `pod.Spec.ServiceAccountName`.

atykhyy commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -178,41 +178,25 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, objRef *v1.ObjectR
-	if err != nil {
-		return "", err.Error(), err
-	}
+		cachedCreds := m.imagePullManager.LookupImagePullCredentials(image, imageRef)
```
OK, I'll rewrite it that way.

stlaz commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -178,41 +178,43 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, objRef *v1.ObjectR
-		m.logIt(objRef, v1.EventTypeNormal, events.PulledImage, logPrefix, msg, klog.Info)
-		return imageRef, msg, nil
-	}
+	lookupPullCredentials := sync.OnceValues(func() ([]credentialprovider.TrackedAuthConfig, error) {
```
Using `sync.OnceValues` here like this would likely result in a continuous memory growth, which is why I avoided it in my original suggestion.

stlaz commented on pkg/kubelet/images/image_manager_test.go:
```diff
@@ -736,7 +741,9 @@ type mockImagePullManagerWithTracking struct {
 
-func (m *mockImagePullManagerWithTracking) MustAttemptImagePull(image, imageRef string, podSecrets []kubeletconfiginternal.ImagePullSecret, podServiceAccount *kubeletconfiginternal.ImagePullServiceAccount) bool {
+func (m *mockImagePullManagerWithTracking) MustAttemptImagePull(image, imageRef string, getPodCredentials pullmanager.GetPodCredentials) (bool, error) {
+	podSecrets, podServiceAccount, _ := getPodCredentials()
```
move right above `m.lastSecrets`, return error if it occured

stlaz commented on pkg/kubelet/images/pullmanager/image_pull_manager_test.go:
```diff
@@ -435,6 +435,7 @@ func TestFileBasedImagePullManager_MustAttemptImagePull(t *testing.T) {
 		pullingFiles       []string
 		expectedPullRecord *kubeletconfiginternal.ImagePulledRecord
 		want               bool
+		wantPodCred        bool
```
nit: `wantPodCredsRequested` would be a more fitting name

atykhyy replied:
Yes, but then there would be a lot of spurious diffs due to `go fmt` whitespace changes. I prefer not to pollute `git blame` if I can help it. But if it's okay then sure I'll change it

atykhyy replied:
The PR went one more size up :|

atykhyy replied:
Just an idea, wouldn't it be better to disregard whitespace changes when computing PR size?

stlaz reviewed (commented):
minor comments, otherwise lgtm

atykhyy commented:
/retest-required

SergeyKanzhelev commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -254,9 +238,47 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, objRef *v1.ObjectR
 }
 
+func (m *imageManager) makeLookupPullCredentialsFunc(image string, pod *v1.Pod, pullSecrets []v1.Secret, podSandboxConfig *runtimeapi.PodSandboxConfig) func() ([]credentialprovider.TrackedAuthConfig, error) {
+	return sync.OnceValues(func() ([]credentialprovider.TrackedAuthConfig, error) {
```
why do we need to wrap it in OnceValues? It is only called once anyways as far as I can tell

stlaz replied:
the return lookup function might be called twice in some cases:
1. JIT during the MustAttemptImagePull
2. just before calling `m.pullImage()`

atykhyy replied:
> Will it retry if the first attempt resulted in `err`?

No. It caches both values.

> Or we want to cache the error

An error in that place shows that something is seriously misconfigured, so it is unlikely that retrying makes sense. A mere failure to lookup credentials from the keyring does not return an error.

atykhyy replied:
> A mere failure to lookup credentials from the keyring does not return an error.

Not because of this PR, I mean. The relevant method's signature does not include `error`.
```go
// pkg/credentialprovider/keyring.go
type DockerKeyring interface {
	Lookup(image string) ([]TrackedAuthConfig, bool)
}
```

> I plan to take it again later during 1.35, please ping if you think it is critical to do earleir

I think this change should be in place when `KubeletEnsureSecretPulledImages` feature gate goes GA and default to enabled, because at that point it will become enabled by some managed (cloud) Kubernetes providers with no option to disable.

stlaz replied:
> An error in that place shows that something is seriously misconfigured, so it is unlikely that retrying makes sense. A mere failure to lookup credentials from the keyring does not return an error.

Besides, I would like to make sure the result of the credential lookup matches what the authz check received in the context of a single `EnsureImageExists()` call.

> I will need to review the sequence you suggested. I plan to take it again later during 1.35, please ping if you think it is critical to do earlier

This likely prevents us from moving the feature to Beta. It's not critical right now, but let's please make sure we pay attention to this PR early in 1.35.

SergeyKanzhelev commented:
@stlaz in the KEP secretly pulled images - when it says "reverification" of secrets, is it really a re-verification that secrets are valid or that they are the same as was used to pull?

Changing the sequence like this changes the assumption on whether secrets should be there. I wanted to ensure that this is exactly what is expected.

SergeyKanzhelev commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -239,8 +213,18 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, objRef *v1.ObjectR
+	}
+
+	if imageRef != "" {
+		if !utilfeature.DefaultFeatureGate.Enabled(features.KubeletEnsureSecretPulledImages) {
```
keeping this code block where it was before will simlpify the review. Took me some time to understand what happened here

stlaz replied:
This is what the code originally looked like before https://github.com/kubernetes/kubernetes/pull/133079. I'd rather it stayed here.

liggitt commented:
> @stlaz in the KEP secretly pulled images - when it says "reverification" of secrets, is it really a re-verification that secrets are valid or that they are the same as was used to pull?

It verifies the pod has access to the *coordinates* of a secret previously used to pull successfully, OR access to identical *content* (hashed) of a credential previously used to pull successfully.

pacoxu reviewed (approved):
/lgtm 
/assign @mrunalp @SergeyKanzhelev

stlaz commented:
/lgtm

atykhyy committed:
kubelet: delay looking up pod image pull credentials until necessary

Some decisions about image pull credential verification can be applied
without reference to image pull credentials: e.g. if the policy is
NeverVerify, or if it is NeverVerifyPreloadedImages and the image is
preloaded, or if is NeverVerifyAllowListedImages and the image is
white-listed. In these cases, there is no need to look up credentials.

haircommander commented:
/lgtm

SergeyKanzhelev reviewed (approved):
/lgtm
/approve
